### ISSUE TITLE
Analyzer type inference only partial if following a negative type test

### ISSUE THREAD

**Issue by lassedamgaard** [dart-lang/sdk#25233]
In the example from WebStorm 11 below the type of `filter` seems to be correctly inferred to be `FacetedFilter` after the initial negative test: the member `facets` which is not defined on the superclass `Filter` is highlighted and I can navigate to it just fine. However, it's also highlighted as an error unless I "reaffirm" that the type is indeed `FacetedFilter`.

![screen shot 2015-12-11 at 11 35 56 2](https://cloud.githubusercontent.com/assets/6475621/11741996/e031f3f8-9ffc-11e5-8b6b-d1561f84407a.png)

**Comment by bwilkerson** [comment]
Unfortunately true. The analyzer is required to report all of the errors and warnings from the specification, and the specification does not include this form of type inference. Hence, even though the analyzer can infer this type for other purposes, it cannot use this information to suppress warnings.

**Comment by lassedamgaard** [comment]
Help me out here: as I mentioned it seems like the type *is* actually inferred, e.g. the member is highlighted and I can jump to its definition. Is this in itself a "bug" (by being inconsistent with the spec) or is it in some perverted way according to spec that the type can be inferred inferred but you for some reason have to warn that the member does not exist anyway even though it does?

**Comment by alexander-doroshko** [comment]
@bwilkerson May be spec should be updated to respect such 100% cases of type inference?

**Comment by bwilkerson** [comment]
> Is this in itself a "bug" (by being inconsistent with the spec) or is it in some perverted way according to spec that the type can be inferred but you for some reason have to warn that the member does not exist anyway even though it does?

I hear your frustration, but I believe that the answer is: neither.

It is clearly an error at runtime to try to invoke a getter on an object that does not define such a getter. Many programmers like to be told before they execute their code that this error will occur, and it is the purpose of the type system to enable that. But no type system is perfect, so every programming language has to make choices about how much type information to try to capture and how strict to be in applying that type information to the detection of errors.

The Dart language specification defines a fairly simple but useful type system. It does allow for a few common cases of type inference, but a choice was made to not try to cover all possible cases of type inference. That does mean that occasionally the specification will specify that a warning should be produced in a place where it would be possible to know that there is no real problem.

However, it is useful for the analyzer to perform more type inference than is included in the specification in order to provide more information during development. It allows us to provide better navigation, better code completion, more accurate refactorings, etc.

That left us with a question: do we produce warnings based on the type information defined by the specification, or from the extra type information we have available. The decision was to produce warnings based on the specification. (In large part, so that analyzer would be consistent with other tools.) There are a lot of implications to changing that decision, so I don't expect it to be changed.

Unfortunately, that decision leads to frustrating situations like the one you ran into. And unfortunately, the only way around it is to do something like you did to tell the analyzer what it already knows. You could also write something like:

    if (filter is FacetedFilter) {
      facets.forEach...
    } else {
      _log.warning...
    }

**Comment by lassedamgaard** [comment]
@bwilkerson I suppose that's a sensible decision. And for the record, what you propose is exactly what I  actually write given this limitation of the spec. However, it's not really satisfactory that this limitation in the type inference rules actually shape my code style, in this case away from using early return for method parameter validation.

My question then is that it would take to get a case such as this included in the spec. I assume the relevant section in the V3 spec is 16.34 which also states that "The rules in the
current specification are deliberately kept simple. It would be upwardly compatible
to refine these rules in the future; such a refinement would accept more code
without warning, but not reject any code now warning-free".

Should I file an issue for this or would it need to go through the DEP process? Seems like a relatively minor addition.

**Comment by eernstg** [comment]
A DEP would be the suitable vehicle for proposing language changes.
However, you might want to know that this kind of extension has been
discussed before, and the underlying problem that makes it hard to draw a
line is the fact that this type of inference is flow-sensitive and hence
undecidable. So there is no upper bound on the number of additional pages
in the spec that you would want to have in order to cover more and more
special cases. You are not alone in your wish to get some of those special
cases included, though. ;-)



On Mon, Dec 14, 2015 at 10:35 AM, lassedamgaard <notifications@github.com>
wrote:

> @bwilkerson <https://github.com/bwilkerson> I suppose that's a sensible
> decision. And for the record, what you propose is exactly what I actually
> write given this limitation of the spec. However, it's not really
> satisfactory that this limitation in the type inference rules actually
> shape my code style, in this case away from using early return for method
> parameter validation.
>
> My question then is that it would take to get a case such as this included
> in the spec. I assume the relevant section in the V3 spec is 16.34 which
> also states that "The rules in the
> current specification are deliberately kept simple. It would be upwardly
> compatible
> to refine these rules in the future; such a refinement would accept more
> code
> without warning, but not reject any code now warning-free".
>
> Should I file an issue for this or would it need to go through the DEP
> process? Seems like a relatively minor addition.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dart-lang/sdk/issues/25233#issuecomment-164389283>.
>



-- 
Erik Ernst  -  Google Danmark ApS
Skt Petri Passage 5, 2 sal, 1165 København K, Denmark
CVR no. 28866984

**Comment by munificent** [comment]
Is this something we could support in strong mode? If so, can I re-open and retarget this bug?

@vsmenon @jmesserly @leafpetersen

**Comment by jmesserly** [comment]
> Is this something we could support in strong mode? If so, can I re-open and retarget this bug?

We don't have anything right now that would do flow sensitive type inference.

My personal opinion: a syntactic form, like pattern matching, is better than trying to extend the type promotion approach.

**Comment by munificent** [comment]
> My personal opinion: a syntactic form, like pattern matching, is better than trying to extend the type promotion approach.

I definitely like pattern matching, but I don't think it's always a good fit. In particular code like the example here that uses checks and bailouts is pretty common and leads to nice-to-read flat code. I'd be sad if people had to add more nesting to please the type inferrer.

I don't think we need much flow sensitive analysis. I believe type promotion today handles:

```dart
if (foo is Bar) {
  // know foo is Bar here
}
```

It may also handle:

```dart
if (foo is! Bar) {
  ...
} else {
  // know foo is Bar here
}
```

I could be wrong, but I don't think it adds much complexity to also support:

```dart
if (foo is! Bar) {
  // definitely exit surrounding block (return, throw, break, continue).
}

// know foo is Bar here
```

Basically, if the then branch definitely exits, then treat the rest of the block like an else branch. Would that work, or is this crazy talk?

**Comment by bwilkerson** [comment]
> Would that work, or is this crazy talk?

Those are not necessarily mutually exclusive. :-)

Analyzer already performs this kind of flow-sensitive analysis for type propagation. It doesn't effect error messages, but it gets us better feedback for code completions, etc.

**Comment by jmesserly** [comment]
We chatted about this, our sense is it's not really a strong mode issue. "strong mode" tries to infer types only when they were omitted. It feels a bit out of the scope. (Doesn't mean it's not worth fixing this issue, just not sure that's the way the fix should be exposed to folks.)

---

Regrading type promotion, the gotchas that will come up:
* mutation of the variable (especially in a closure) can cause action-at-a-distance that disables promotion
* unless you're downcasting, it won't promote. So "S x; if (x is T) { ... }", only works if T <: S
* in the common case where "S" is dynamic, you can't really tell if type promotion is working, because messages are suppressed anyway

Those can be pretty surprising, and can happen as code evolves.

We could try and make type promotion more reliable, but I don't think it can ever be 100% due to the mutate-in-closure issue. Fixing the downcasting issue requires intersection types, which Analyzer doesn't currently have a way to track. Maybe fixing the early-exit issue isn't too bad. I'm not very familiar with the existing propagated type inference, so not sure how hard it would be.
